### PR TITLE
feat(stdlib): Add `toIntegerRatio` and `fromIntegerRatio`  to Rational Library

### LINKED ISSUE #1672: Add new Rational functions
One last small comment. The only things I feel are missing from the initial library are `numerator` and `denominator`.

Things that can come later (and we can likely convert this comment to an issue) would be:
- `toIntegerRatio`, to get a tuple of numerator and denominator
- `fromIntegerRatio`, to construct a Rational from two integers
- `toProperFraction`, to get a tuple of an integer and a rational < 1
- `toFloat`
- `inverse`, to invert the rational
- `rationalize`, to get a simpler rational (maybe with an optional epsilon argument)—Python has `limit_denominator` which I maybe like a bit more, you essentially give it a target denominator and it gives you back a rational with a denominator of at most what you requested

_Originally posted by @ospencer in https://github.com/grain-lang/grain/pull/1603#pullrequestreview-1294576997_

### PR DESCRIPTION
**Pull request by spotandjake**
This adds `toIntegerRation` and `fromIntegerRation`, to the `rational` library.

This doesnt close #1672, but it adds a few of the functions for it.

### CODE REVIEW

File: stdlib/rational.gr [hunk 1]
```diff
@@ -23,13 +23,39 @@ from Numbers use {
...
+ * Makes a rational from the numerator and denominator
+ *
+ * @param numerator: The numerator
+ * @param denominator: The denominator
+ * @returns The built rational
+ *
+ * @since v0.6.0
+ */
+provide let fromIntegerRatio = (numerator: Number, denominator: Number) => {
+  fromNumber(numerator / denominator)
```
Comment (alex-snezhko) [inline stdlib/rational.gr:49]: I think this should throw an exception if the inputs are not integers

File: stdlib/rational.gr [hunk 2]
```diff
@@ -19,17 +19,54 @@ from Numbers use {
...
+/**
+ * Gets the numerator and denominator of the rational as a tuple
+ *
+ * @param x: The rational to convert
+ * @returns The numerator and denominator of the rational
+ *
+ * @since v0.6.0
+ */
+@unsafe
+provide let toIntegerRatio = (x: Rational) => {
```
[the remark this answers is not in the document]
Comment (spotandjake) [reply stdlib/rational.gr:36]: nope

File: stdlib/rational.gr [hunk 3]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
...
   cmpRationals,
+  isInteger,
 }
 
 provide { fromNumber, toNumber, numerator, denominator }
 
+/**
+ * Gets the numerator and denominator of the rational as a tuple.
+ *
+ * @param x: The rational to convert
```
Comment (ospencer) [inline stdlib/rational.gr:30]: Something like this maybe?
```suggestion
 * @param x: The rational to split
```

File: stdlib/rational.gr [hunk 4]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
...
+provide let toIntegerRatio = x => {
+  (numerator(x), denominator(x))
+}
+
+/**
+ * Creates a rational from a numerator and denominator.
+ *
+ * @param numerator: The numerator
+ * @param denominator: The denominator
+ * @returns The built rational
```
Comment (ospencer) [inline stdlib/rational.gr:44]:
```suggestion
 * @returns The newly constructed rational
```

File: stdlib/rational.gr [hunk 5]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
   timesDivideRational,
   rationalsEqual,
   cmpRationals,
+  isInteger,
 }
 
 provide { fromNumber, toNumber, numerator, denominator }
 
+/**
+ * Gets the numerator and denominator of the rational as a tuple.
```
Comment (phated) [inline stdlib/rational.gr:28]:
```suggestion
 * Gets the numerator and denominator of the rational.
```
The return type is specified in the table

File: stdlib/rational.gr [hunk 6]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
...
+}
+
+/**
+ * Creates a rational from a numerator and denominator.
+ *
+ * @param numerator: The numerator
+ * @param denominator: The denominator
+ * @returns The newly constructed rational
+ *
+ * @throws InvalidArgument(String): If the numerator or denominator are not integers
```
Comment (phated) [inline stdlib/rational.gr:46]: Split this into two throws "if the numerator is not an integer" and "if the denominator is not an integer"

File: stdlib/rational.gr [hunk 7]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
...
+provide let toIntegerRatio = x => {
+  (numerator(x), denominator(x))
+}
+
+/**
+ * Creates a rational from a numerator and denominator.
+ *
+ * @param numerator: The numerator
+ * @param denominator: The denominator
+ * @returns The newly constructed rational
```
Comment (phated) [inline stdlib/rational.gr:44]:
```suggestion
 * @returns The reduced rational 
```

File: stdlib/rational.gr [hunk 8]
```diff
@@ -19,17 +19,53 @@ from Numbers use {
...
+ */
+@unsafe
+provide let fromIntegerRatio = (numerator: Number, denominator: Number) => {
+  let numeratorptr = WasmI32.fromGrain(numerator)
+  let denominatorptr = WasmI32.fromGrain(denominator)
+  if (!isInteger(numeratorptr) || !isInteger(denominatorptr)) {
+    throw Exception.InvalidArgument(
+      "Numerator and denominator must be integers"
+    )
+  }
```
Comment (phated) [inline stdlib/rational.gr:58]: Separate this throw saying which one is the problem

**Review by phated — CHANGES_REQUESTED** [review]
Needs a few more changes.

**Review by phated — APPROVED** [review]
Excellent!

**Pull request merged**